go/consensus/cometbft/apps/vault/transactions.go

```diff
@@ -0,0 +1,277 @@
+			Nonce:  authAction.Nonce,
+			Action: authAction.Action,
+		}
+	default:
+		return err
+	}
+
+	ctx.EmitEvent(api.NewEventBuilder(app.Name()).TypedAttribute(&vault.ActionSubmittedEvent{
+		Submitter: ctx.CallerAddress(),
+		Vault:     authAction.Vault,
+		Nonce:     authAction.Nonce,
+	}))
+
+	// Update list of authorizers.
+	if !pendingAction.ContainsAuthorizationFrom(ctx.CallerAddress()) {
+		pendingAction.AuthorizedBy = append(pendingAction.AuthorizedBy, ctx.CallerAddress())
+		if err = state.SetPendingAction(ctx, authAction.Vault, pendingAction); err != nil {
+			return err
+		}
+	}
+
+	// Check if action has become executable.
+	var canExecute bool
+	for _, auth := range pendingAction.Action.Authorities(vlt) {
+		if auth.Verify(pendingAction.AuthorizedBy) {
+			canExecute = true
+			break
+		}
+	}
+	if !canExecute {
+		ctx.Commit()
+		return nil
+	}
+
+	// Execute action.
+	evExec := &vault.ActionExecutedEvent{
+		Vault: authAction.Vault,
+		Nonce: authAction.Nonce,
+	}
+	err = app.executeAction(ctx, vlt, &pendingAction.Action)
+	switch {
+	case api.IsUnavailableStateError(err):
+		// Propagate state unavailability errors.
+		return err
+	default:
+		// Record other errors (or success) in the execution event.
+		evExec.Result.Module, evExec.Result.Code = errors.Code(err)
+
+		ctx.Logger().Debug("vault executed action",
+			"err", err,
+			"vault", authAction.Vault,
+			"nonce", authAction.Nonce,
+			"action", pendingAction.Action,
+		)
+	}
+
+	ctx.EmitEvent(api.NewEventBuilder(app.Name()).TypedAttribute(evExec))
+
+	// Remove pending action as it has been executed.
+	if err = state.RemovePendingAction(ctx, authAction.Vault, authAction.Nonce); err != nil {
+		return err
+	}
+
+	vlt.Nonce++
+	if err = state.SetVault(ctx, vlt); err != nil {
+		return err
+	}
+
+	ctx.Commit()
+
+	return nil
+}
+
+func (app *vaultApplication) cancelAction(ctx *api.Context, cancelAction *vault.CancelAction) error {
+	// Validate arguments.
+	if err := cancelAction.Validate(); err != nil {
+		return fmt.Errorf("%w: %w", vault.ErrInvalidArgument, err)
+	}
+
+	// Ensure vault exists.
+	state := vaultState.NewMutableState(ctx.State())
+	vlt, err := state.Vault(ctx, cancelAction.Vault)
+	if err != nil {
+		return err
+	}
+
+	// Validate action nonce. Currently queuing multiple future actions is not allowed.
+	if cancelAction.Nonce != vlt.Nonce {
+		return vault.ErrInvalidNonce
+	}
+
+	// Before we know what the canceled action is, we can only check that the caller is part of at
+	// least one of the vault's authorities. Later, we will check against the action authority.
+	if !vlt.AuthoritiesContain(ctx.CallerAddress()) {
+		return vault.ErrForbidden
+	}
+
+	if ctx.IsCheckOnly() {
+		return nil
+	}
+
+	// Charge gas for this transaction.
+	params, err := state.ConsensusParameters(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to fetch consensus parameters: %w", err)
+	}
+	if err = ctx.Gas().UseGas(1, vault.GasOpAuthorizeAction, params.GasCosts); err != nil {
```

Comment:
```suggestion
	if err = ctx.Gas().UseGas(1, vault.GasOpCancelAction, params.GasCosts); err != nil {
```
Cancel actions could also charge lower gas.